pkg/eventingtls/servermanager.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2023 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventingtls
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"knative.dev/eventing/pkg/apis/feature"
+	"knative.dev/eventing/pkg/kncloudevents"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/logging"
+)
+
+// ServerManager is intended to be used to manage HTTP and HTTPS servers for a component.
+// It relies on the `transport-encryption` feature flag to determine which server(s) should be accepting requests.
+// If a server shouldn't be accepting requests, ServerManager will update that server's handler to respond with a 404
+//
+// disabled: only http server
+// permissive: both http and https servers
+// strict: only https server
+type ServerManager struct {
+	httpReceiver  *kncloudevents.HTTPMessageReceiver
+	httpsReceiver *kncloudevents.HTTPMessageReceiver
+	handler       http.Handler
+	cmw           configmap.Watcher
+	featureStore  *feature.Store
+}
+
+func NewServerManager(ctx context.Context, httpReceiver, httpsReceiver *kncloudevents.HTTPMessageReceiver, handler http.Handler, cmw configmap.Watcher) (*ServerManager, error) {
+	if httpReceiver == nil || httpsReceiver == nil {
+		return nil, fmt.Errorf("message receiver not provided")
+	}
+
+	featureStore := feature.NewStore(logging.FromContext(ctx).Named("feature-config-store"))
+	featureStore.WatchConfigs(cmw)
+
+	return &ServerManager{
+		httpReceiver:  httpReceiver,
+		httpsReceiver: httpsReceiver,
+		handler:       handler,
+		cmw:           cmw,
+		featureStore:  featureStore,
+	}, nil
+}
+
+// Blocking call. Starts the 2 servers
+func (s *ServerManager) StartServers(ctx context.Context) (httpError, httpsError error) {
+	// start servers
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		httpError = s.httpReceiver.StartListen(ctx, s.httpHandler())
+	}()
+
+	go func() {
+		defer wg.Done()
+		httpsError = s.httpsReceiver.StartListen(ctx, s.httpsHandler())
+	}()
+
+	wg.Wait()
+	return
```

Comment:
Not sure on this, but wouldn't it be impossible to stop the other server if one fails, as this is blocking?

Reply:
it waits until the given list is processed/empty

Reply:
With "list" you mean the waitgroup? I am thinking about the scenario, when one receiver dies and the other one is still running. Then this method still waits until the other receiver finishes/dies and from a caller perspective (caller of `StartServer`) there is no way to react to one failed receiver (IIUC).
But maybe I simply don't get it at the moment :shrug: :smile:

Reply:
I like the idea of failing fast when on of them fails we return an error (maybe something like a select statement on an error channel)